File: yaml/tests/R/runner.R
# Execute a specification, R implementation.
#
# The counterpart of tests/python/runner.py. Same phases, same rules, and it must
# serialize to byte-identical output.

ODM_CONTEXT <- c("StudyOID", "SubjectKey", "StudyEventOID",
                 "StudyEventRepeatKey", "ItemGroupOID", "ItemGroupRepeatKey")

HANDLED_STAGES <- c("bind", "join", "operation", "convert", "final")

# Every registered name must be dispatchable. The counterpart of
# check_implemented() in tests/python/test_fixtures.py.
check_implemented <- function(d = load_design()) {
  errs <- character()
  tables <- list(scalar = SCALAR_OPS, window = WINDOW_OPS, aggregate = AGGREGATE_OPS)
  dispatch <- c(names(SCALAR_OPS), names(WINDOW_OPS), names(AGGREGATE_OPS))
  for (nm in names(d$ops)) {
    if (!(nm %in% dispatch)) {
      errs <- c(errs, sprintf("operation '%s' is registered but not implemented", nm))
      next
    }
    if (!(nm %in% names(tables[[d$ops[[nm]]$kind]]))) {
      errs <- c(errs, sprintf("operation '%s' declares kind '%s' but is implemented elsewhere",
                              nm, d$ops[[nm]]$kind))
    }
  }
  for (nm in names(d$exc)) {
    if (!(d$exc[[nm]]$stage %in% HANDLED_STAGES)) {
      errs <- c(errs, sprintf("exception '%s' binds to stage '%s', which the engine ignores",
                              nm, d$exc[[nm]]$stage))
    }
  }
  errs
}

read_csv_chr <- function(path) {
  df <- utils::read.csv(path, colClasses = "character", check.names = FALSE,
                        na.strings = character(0))
  lapply(seq_len(nrow(df)), function(i) as.list(df[i, , drop = FALSE]))
}

exception_map <- function(dv, exc_reg) {
  out <- list()
  for (ex in as_list_of(dv$exception)) {
    nm <- names(ex)[1]
    out[[exc_reg[[nm]]$stage]] <- list(name = nm, args = ex[[1]] %||% list())
  }
  out
}

run_spec <- function(path, d = load_design()) {
  spec <- read_yaml_1_2(path)
  base <- spec$base
  keys <- unlist(spec$keys %||% list())
  coltype <- stats::setNames(lapply(spec$columns, function(c) c$type),
                             vapply(spec$columns, function(c) c$name, ""))
  order_cols <- vapply(spec$columns, function(c) c$name, "")
  dir <- dirname(path)
  data <- lapply(spec$datasets, function(rel) read_csv_chr(file.path(dir, rel)))

  rows <- list()   # each: list(src=, driver=, tmpl=, out=list())

  templates <- spec$rows %||% list()
  if (length(templates)) {
    for (ti in seq_along(templates)) {
      t <- templates[[ti]]
      driver <- t$dataset %||% base
      for (rec in data[[driver]]) {
        if (!is.null(t$filter)) {
          look <- rec_lookup(rec, driver)
          if (!evaluate_predicate(t$filter, look)) next
        }
        rows[[length(rows) + 1L]] <- list(src = rec, driver = driver, tmpl = ti, out = list())
      }
    }
  } else {
    if (is.null(base)) stop("no rows entry and no base (R001)")
    for (rec in data[[base]]) {
      rows[[length(rows) + 1L]] <- list(src = rec, driver = base, tmpl = NA, out = list())
    }
  }

  ctx <- new.env(parent = emptyenv())
  ctx$data <- data
  ctx$rows <- rows
  ctx$keys <- keys
  ctx$base <- base
  ctx$coltype <- coltype
  ctx$spec <- spec
  ctx$ops <- d$ops
  ctx$exc <- d$exc

  # phase 1: row-template derivations
  for (ti in seq_along(templates)) {
    t <- templates[[ti]]
    idx <- which(vapply(ctx$rows, function(r) identical(r$tmpl, ti), logical(1)))
    driver <- t$dataset %||% base
    for (nm in toposort_defs(t$derivations, order_cols, ctx, driver)) {
      vals <- evaluate_derivation(nm, t$derivations[[nm]], idx, ctx, driver)
      for (k in seq_along(idx)) ctx$rows[[idx[k]]]$out[nm] <- list(vals[[k]])
    }
  }

  # phase 2: column derivations
  coldefs <- list()
  for (c in spec$columns) if (!is.null(c$derivation)) coldefs[[c$name]] <- c$derivation
  all_idx <- seq_along(ctx$rows)
  for (nm in toposort_defs(coldefs, order_cols, ctx, base)) {
    vals <- evaluate_derivation(nm, coldefs[[nm]], all_idx, ctx, base)
    for (k in all_idx) ctx$rows[[k]]$out[nm] <- list(vals[[k]])
  }

  # R005: order output rows by keys
  if (length(keys)) {
    kv <- lapply(keys, function(k) lapply(ctx$rows, function(r) r$out[[k]]))
    args <- unlist(lapply(kv, function(vals) {
      list(vapply(vals, function(v) ord_key(v)[1], 0),
           vapply(vals, function(v) ord_key(v)[2], 0),
           vapply(vals, function(v) if (is_missing(v)) "" else as.character(v), ""))
    }), recursive = FALSE)
    ctx$rows <- ctx$rows[do.call(order, args)]
  }

  lapply(ctx$rows, function(r) {
    stats::setNames(lapply(order_cols, function(c) serialize_value(r$out[[c]])), order_cols)
  })
}

rec_lookup <- function(rec, ds) {
  function(name) {
    key <- if (startsWith(name, paste0(ds, "."))) substring(name, nchar(ds) + 2L) else name
    rec[[key]]
  }
}

odm_item <- function(ctx, ds, item, src) {
  ctxkeys <- intersect(ODM_CONTEXT, names(src))
  hits <- Filter(function(r) {
    identical(r[["ItemOID"]], item) &&
      all(vapply(ctxkeys, function(k) identical(r[[k]], src[[k]]), logical(1)))
  }, ctx$data[[ds]])
  if (length(hits) > 1L) stop(sprintf("%s.%s: %d records match the context", ds, item, length(hits)))
  if (!length(hits)) return(NULL)
  hits[[1]][["Value"]]
}

joined_value <- function(ctx, ds, col, ri, dv) {
  right <- ctx$data[[ds]]
  if (!is.null(dv$filter)) {
    right <- Filter(function(r) evaluate_predicate(dv$filter, rec_lookup(r, ds)), right)
  }
  applicable <- if (length(right)) intersect(ctx$keys, names(right[[1]])) else character()
  if (!length(applicable)) stop(sprintf("%s.%s: no applicable keys (R003)", ds, col))

  ops <- as_list_of(dv$operations)
  reducing <- length(ops) && identical(ctx$ops[[names(ops[[1]])[1]]]$kind, "aggregate")
  want <- lapply(applicable, function(k) ctx$rows[[ri]]$out[[k]])
  hits <- Filter(function(r) {
    all(vapply(seq_along(applicable), function(j) {
      identical(as.character(r[[applicable[j]]]), as.character(want[[j]] %||% ""))
    }, logical(1)))
  }, right)

  if (reducing) {
    fn <- AGGREGATE_OPS[[names(ops[[1]])[1]]]
    return(fn(lapply(hits, function(h) h[[col]])))
  }
  if (length(hits) > 1L) {
    ex <- exception_map(dv, ctx$exc)$join
    if (is.null(ex) || !identical(ex$name, "multiple_matches")) {
      stop(sprintf("%s.%s: %d right-side matches (R003)", ds, col, length(hits)))
    }
    ocol <- sub("^.*\\.", "", if (is.list(ex$args$order_by)) ex$args$order_by$source else ex$args$order_by)
    idx <- sort_index(lapply(hits, function(h) h[[ocol]]))
    hits <- list(hits[[if (identical(ex$args$keep, "last")) idx[length(idx)] else idx[1]]])
  }
  if (!length(hits)) return(NULL)
  hits[[1]][[col]]
}

resolve_ref <- function(ctx, ref, ri, dv) {
  if (!grepl(".", ref, fixed = TRUE)) {
    r <- ctx$rows[[ri]]
    if (!(ref %in% names(r$out))) stop(sprintf("%s is not available yet (R001)", ref))
    return(r$out[[ref]])
  }
  parts <- strsplit(ref, ".", fixed = TRUE)[[1]]
  ds <- parts[1]; rest <- paste(parts[-1], collapse = ".")
  r <- ctx$rows[[ri]]
  if (identical(ds, r$driver)) {
    if (rest %in% names(r$src)) return(r$src[[rest]])
    return(odm_item(ctx, ds, rest, r$src))
  }
  first <- if (length(ctx$data[[ds]])) names(ctx$data[[ds]][[1]]) else character()
  if (rest %in% first) return(joined_value(ctx, ds, rest, ri, dv))
  odm_item(ctx, ds, rest, r$src)
}

resolve_args <- function(ctx, raw, ri, dv) {
  if (is.list(raw) && identical(names(raw), "source")) return(resolve_ref(ctx, raw$source, ri, dv))
  if (is.list(raw)) return(lapply(raw, function(v) resolve_args(ctx, v, ri, dv)))
  raw
}

deps_of <- function(dv, ctx, driver) {
  out <- character()
  src <- dv$source
  if (!is.null(src) && !grepl(".", src, fixed = TRUE)) out <- c(out, src)
  out <- c(out, unlist(dv$group_by %||% list()))
  if (!is.null(dv$where)) {
    v <- predicate_vars(dv$where)
    out <- c(out, v[!grepl(".", v, fixed = TRUE)])
  }
  if (!is.null(src) && grepl(".", src, fixed = TRUE) &&
      !identical(sub("\\..*$", "", src), driver %||% "")) {
    out <- c(out, ctx$keys)
  }
  collect <- function(x) {
    if (is.list(x)) {
      if (identical(names(x), "source") && !grepl(".", x$source, fixed = TRUE)) {
        out <<- c(out, x$source)
      } else {
        for (nm in names(x)) {
          if (identical(nm, "when") && is.character(x[[nm]])) {
            v <- predicate_vars(x[[nm]]); out <<- c(out, v[!grepl(".", v, fixed = TRUE)])
          } else collect(x[[nm]])
        }
        if (is.null(names(x))) for (e in x) collect(e)
      }
    }
  }
  for (op in as_list_of(dv$operations)) collect(op[[1]])
  for (ex in as_list_of(dv$exception)) collect(ex[[1]])
  unique(out)
}

toposort_defs <- function(defs, order_cols, ctx, driver) {
  pending <- names(defs); done <- character()
  pos <- stats::setNames(seq_along(order_cols), order_cols)
  while (length(pending)) {
    ready <- pending[vapply(pending, function(n) {
      d <- setdiff(intersect(deps_of(defs[[n]], ctx, driver), pending), n)
      !length(d)
    }, logical(1))]
    if (!length(ready)) stop(sprintf("dependency cycle among %s (R001)",
                                     paste(sort(pending), collapse = ", ")))
    ready <- ready[order(pos[ready])]
    done <- c(done, ready); pending <- setdiff(pending, ready)
  }
  done
}

evaluate_derivation <- function(name, dv, idx, ctx, driver) {
  exc <- exception_map(dv, ctx$exc)

  if (!is.null(dv$where)) {
    keep <- idx[vapply(idx, function(ri) {
      evaluate_predicate(dv$where, function(n) ctx$rows[[ri]]$out[[n]])
    }, logical(1))]
    sub_dv <- dv; sub_dv$where <- NULL
    sub <- evaluate_derivation(name, sub_dv, keep, ctx, driver)
    out <- vector("list", length(idx))
    for (k in seq_along(keep)) out[which(idx == keep[k])] <- list(sub[[k]])
    return(out)
  }

  vals <- lapply(idx, function(ri) {
    v <- tryCatch({
      if (!is.null(dv$literal)) dv$literal
      else if (!is.null(dv$source)) resolve_ref(ctx, dv$source, ri, dv)
      else NULL
    }, error = function(e) if (!is.null(exc$bind)) exc$bind$args$default else stop(e))
    if (is_missing(v) && !is.null(exc$bind)) v <- exc$bind$args$default
    v
  })

  for (op in as_list_of(dv$operations)) {
    opname <- names(op)[1]; raw <- op[[1]] %||% list()
    kind <- ctx$ops[[opname]]$kind
    if (identical(kind, "aggregate")) next            # applied during R003 reduction
    if (identical(kind, "window")) {
      vals <- apply_window(opname, raw, idx, vals, ctx, dv)
      next
    }
    fn <- SCALAR_OPS[[opname]]
    if (is.null(fn)) stop(sprintf("%s is registered but not implemented", opname))
    vals <- lapply(seq_along(idx), function(k) {
      ri <- idx[k]
      args <- resolve_args(ctx, raw, ri, dv)
      if (identical(opname, "call")) {
        args <- list(function_name = raw[["function"]],
                     args = resolve_args(ctx, raw$args %||% list(), ri, dv))
      }
      octx <- list(data = ctx$data, row_lookup = function(n) ctx$rows[[ri]]$out[[n]])
      tryCatch(fn(vals[[k]], args, octx), error = function(e) {
        if (identical(conditionMessage(e), UNMAPPED)) {
          if (is.null(exc$operation)) {
            stop(sprintf("%s: %s could not map a value and no unmapped exception is declared (R008)",
                         name, opname))
          }
          return(exc$operation$args$default)
        }
        stop(e)
      })
    })
  }

  conv <- lapply(vals, function(v) {
    tryCatch(convert_value(v, ctx$coltype[[name]]), error = function(e) {
      if (is.null(exc$convert)) stop(e)
      convert_value(exc$convert$args$default, ctx$coltype[[name]])
    })
  })

  if (!is.null(exc$final)) {
    conv <- lapply(seq_along(idx), function(k) {
      ri <- idx[k]; cur <- conv[[k]]
      for (rule in exc$final$args$rules) {
        look <- function(n) if (identical(n, name)) cur else ctx$rows[[ri]]$out[[n]]
        if (evaluate_predicate(rule$when, look)) return(convert_value(rule$value, ctx$coltype[[name]]))
      }
      cur
    })
  }
  conv
}

apply_window <- function(opname, raw, idx, vals, ctx, dv) {
  gb <- unlist(dv$group_by %||% list())
  gkey <- vapply(idx, function(ri) {
    paste(vapply(gb, function(k) as.character(ctx$rows[[ri]]$out[[k]] %||% ""), ""), collapse = "\r")
  }, "")
  out <- vector("list", length(idx))
  for (g in unique(gkey)) {
    sel <- which(gkey == g)
    sub_idx <- idx[sel]
    args <- list()
    for (nm in names(raw)) {
      spec <- raw[[nm]]
      if (is.list(spec) && is.null(names(spec))) {
        args[[paste0(nm, "_per_row")]] <- lapply(sub_idx, function(ri) {
          lapply(spec, function(s) resolve_args(ctx, s, ri, dv))
        })
      } else {
        args[[paste0(nm, "_per_row")]] <- lapply(sub_idx, function(ri) resolve_args(ctx, spec, ri, dv))
      }
    }
    res <- WINDOW_OPS[[opname]](as.list(sub_idx), args,
                               list(data = ctx$data, row_lookup = function(n) NULL))
    for (k in seq_along(sel)) out[sel[k]] <- list(res[[k]])
  }
  out
}

File: yaml/tests/python/runner.py
"""Execute a specification end to end and compare against its expected CSV."""
from __future__ import annotations

import csv
from pathlib import Path

import yaml

import engine
from engine import (
    AGGREGATE_OPS,
    ODM_CONTEXT,
    SCALAR_OPS,
    WINDOW_OPS,
    DerivationError,
    Unmapped,
    convert,
    evaluate_predicate,
    predicate_vars,
    serialize,
)
from validate import EXC, OPS


def _as_list(v):
    return [] if v is None else (v if isinstance(v, list) else [v])


def _exception_map(deriv):
    """Exception name by R008 stage for one derivation."""
    out = {}
    for ex in _as_list(deriv.get("exception")):
        name, args = next(iter(ex.items()))
        out[EXC[name]["stage"]] = (name, args or {})
    return out


class Row:
    __slots__ = ("src", "driver", "template", "out")

    def __init__(self, src, driver, template):
        self.src = src
        self.driver = driver
        self.template = template
        self.out = {}


class Spec:
    def __init__(self, path):
        self.path = Path(path)
        self.spec = yaml.safe_load(self.path.read_text())
        self.base = self.spec.get("base")
        self.keys = self.spec.get("keys", [])
        self.columns = self.spec["columns"]
        self.coltype = {c["name"]: c["type"] for c in self.columns}
        self.order = [c["name"] for c in self.columns]
        self.data = {
            name: list(csv.DictReader(open(self.path.parent / rel)))
            for name, rel in self.spec["datasets"].items()
        }
        self.rows = []
        self.row_lookup = lambda name: None

    # -- reference resolution ------------------------------------------

    def _odm_item(self, ds, item, src):
        """R002 contextual reference: an ItemOID resolved within the current
        study, subject, event, item-group and repeat-key context."""
        ctx = {k: src.get(k) for k in ODM_CONTEXT if k in src}
        hits = [
            r
            for r in self.data[ds]
            if r.get("ItemOID") == item and all(r.get(k) == v for k, v in ctx.items())
        ]
        if len(hits) > 1:
            raise DerivationError(f"{ds}.{item}: {len(hits)} records match the context")
        return hits[0].get("Value") if hits else None

    def _joined(self, ds, col, row, deriv):
        """R003 left join on applicable keys, with optional right-side reduction."""
        right = self.data[ds]
        if deriv.get("filter"):
            right = [
                r for r in right if evaluate_predicate(deriv["filter"], _rec_lookup(r, ds))
            ]
        applicable = [k for k in self.keys if right and k in right[0]]
        if not applicable:
            raise DerivationError(f"{ds}.{col}: no applicable keys (R003)")

        ops = _as_list(deriv.get("operations"))
        reducing = ops and OPS[next(iter(ops[0]))]["kind"] == "aggregate"
        want = tuple(row.out.get(k) for k in applicable)
        hits = [r for r in right if tuple(r.get(k) for k in applicable) == want]

        if reducing:
            fn = AGGREGATE_OPS[next(iter(ops[0]))]
            return fn([h.get(col) for h in hits])
        if len(hits) > 1:
            stage = _exception_map(deriv).get("join")
            if not stage or stage[0] != "multiple_matches":
                raise DerivationError(f"{ds}.{col}: {len(hits)} right-side matches (R003)")
            _, args = stage
            ordered = sorted(hits, key=lambda h: str(h.get(_var(args["order_by"]).split(".")[-1])))
            hits = [ordered[-1] if args["keep"] == "last" else ordered[0]]
        return hits[0].get(col) if hits else None

    def resolve(self, ref, row, deriv):
        if "." not in ref:
            if ref not in row.out:
                raise DerivationError(f"{ref} is not available yet (R001)")
            return row.out[ref]
        ds, rest = ref.split(".", 1)
        if ds == row.driver:
            if rest in row.src:
                return row.src[rest]
            return self._odm_item(ds, rest, row.src)
        if rest in (self.data[ds][0] if self.data[ds] else {}):
            return self._joined(ds, rest, row, deriv)
        return self._odm_item(ds, rest, row.src)

    # -- dependencies ---------------------------------------------------

    def deps(self, deriv, driver):
        out = set()
        src = deriv.get("source")
        if src and "." not in str(src):
            out.add(src)
        for k in deriv.get("group_by") or []:
            out.add(k)
        if src and "." in str(src) and str(src).split(".")[0] != driver:
            # an R003 join needs its applicable keys first
            out.update(self.keys)
        for op in _as_list(deriv.get("operations")):
            out |= _arg_deps(next(iter(op.values())) or {})
        for ex in _as_list(deriv.get("exception")):
            out |= _arg_deps(next(iter(ex.values())) or {})
        return out

    # -- evaluation -----------------------------------------------------

    def _args(self, raw, row, deriv):
        if isinstance(raw, dict) and set(raw) == {"source"}:
            return self.resolve(raw["source"], row, deriv)
        if isinstance(raw, dict):
            return {k: self._args(v, row, deriv) for k, v in raw.items()}
        if isinstance(raw, list):
            return [self._args(v, row, deriv) for v in raw]
        return raw

    def evaluate(self, name, deriv, rows):
        exc = _exception_map(deriv)
        vals = []
        for row in rows:
            try:
                if "literal" in deriv:
                    v = deriv["literal"]
                elif "source" in deriv:
                    v = self.resolve(deriv["source"], row, deriv)
                    if v in (None, "") and "bind" in exc:
                        v = exc["bind"][1]["default"]
                else:
                    v = None
            except DerivationError:
                if "bind" not in exc:
                    raise
                v = exc["bind"][1]["default"]
            vals.append(v)

        ops = _as_list(deriv.get("operations"))
        for i, op in enumerate(ops):
            opname, raw = next(iter(op.items()))
            raw = raw or {}
            kind = OPS[opname]["kind"]
            if kind == "aggregate":
                continue  # applied during right-side reduction (R003)
            if kind == "window":
                vals = self._window(opname, raw, rows, vals, deriv)
                continue
            fn = SCALAR_OPS.get(opname)
            if fn is None:
                raise DerivationError(f"{opname} is registered but not implemented")
            out = []
            for row, v in zip(rows, vals):
                self.row_lookup = lambda n, r=row: r.out.get(n)
                engine_ctx = _Ctx(self.data, self.row_lookup)
                try:
                    out.append(fn(v, self._args(raw, row, deriv), engine_ctx))
                except Unmapped:
                    if "operation" not in exc:
                        raise DerivationError(
                            f"{name}: {opname} could not map {v!r} and no unmapped "
                            f"exception is declared (R008)"
                        )
                    out.append(exc["operation"][1]["default"])
            vals = out

        conv = []
        for v in vals:
            try:
                conv.append(convert(v, self.coltype[name]))
            except DerivationError:
                if "convert" not in exc:
                    raise
                conv.append(convert(exc["convert"][1]["default"], self.coltype[name]))
        return conv

    def _window(self, opname, raw, rows, vals, deriv):
        groups = {}
        gb = deriv.get("group_by") or []
        for i, row in enumerate(rows):
            groups.setdefault(tuple(row.out.get(k) for k in gb), []).append(i)
        out = [None] * len(rows)
        for idx in groups.values():
            sub = [rows[i] for i in idx]
            args = {}
            for key, spec in raw.items():
                if isinstance(spec, list):
                    args[f"{key}_per_row"] = [
                        [self._args(s, r, deriv) for s in spec] for r in sub
                    ]
                else:
                    args[f"{key}_per_row"] = [self._args(spec, r, deriv) for r in sub]
            res = WINDOW_OPS[opname](sub, args, _Ctx(self.data, self.row_lookup))
            for pos, i in enumerate(idx):
                out[i] = res[pos]
        return out

    # -- driver ---------------------------------------------------------

    def run(self):
        templates = self.spec.get("rows") or []
        if templates:
            for t in templates:
                driver = t.get("dataset", self.base)
                for rec in self.data[driver]:
                    if t.get("filter") and not evaluate_predicate(
                        t["filter"], _rec_lookup(rec, driver)
                    ):
                        continue
                    self.rows.append(Row(rec, driver, t))
        else:
            if not self.base:
                raise DerivationError("no rows entry and no base (R001)")
            for rec in self.data[self.base]:
                self.rows.append(Row(rec, self.base, None))

        # phase 1: row-template derivations, per template, dependency ordered
        for t in templates:
            sub = [r for r in self.rows if r.template is t]
            driver = t.get("dataset", self.base)
            for name in _toposort(
                t["derivations"], self.order, lambda d: self.deps(d, driver)
            ):
                for r, v in zip(sub, self.evaluate(name, t["derivations"][name], sub)):
                    r.out[name] = v

        # phase 2: column derivations across all rows
        coldefs = {c["name"]: c["derivation"] for c in self.columns if "derivation" in c}
        for name in _toposort(
            coldefs, self.order, lambda d: self.deps(d, self.base)
        ):
            for r, v in zip(self.rows, self.evaluate(name, coldefs[name], self.rows)):
                r.out[name] = v

        self.rows.sort(key=lambda r: tuple(_ord(r.out.get(k)) for k in self.keys))
        return [{c: serialize(r.out.get(c)) for c in self.order} for r in self.rows]

    def expected(self):
        exp = self.path.parent / "expected"
        target = next(exp.glob("*.csv"))
        with open(target) as fh:
            return [dict(r) for r in csv.DictReader(fh)]


class _Ctx:
    def __init__(self, data, row_lookup):
        self.data = data
        self.row_lookup = row_lookup


def _ord(v):
    """Sort key for R005 output ordering: missing last, numbers before text."""
    if v is None or v == "":
        return (2, 0.0, "")
    if isinstance(v, (int, float)) and not isinstance(v, bool):
        return (0, float(v), "")
    return (1, 0.0, str(v))


def _var(a):
    return a["source"] if isinstance(a, dict) and "source" in a else str(a)


def _rec_lookup(rec, ds):
    def look(name):
        return rec.get(name.split(".", 1)[1] if name.startswith(ds + ".") else name)

    return look


def _arg_deps(raw):
    out = set()
    if isinstance(raw, dict):
        if set(raw) == {"source"} and "." not in str(raw["source"]):
            out.add(raw["source"])
        else:
            for k, v in raw.items():
                if k == "when" and isinstance(v, str):
                    out |= {x for x in predicate_vars(v) if "." not in x}
                else:
                    out |= _arg_deps(v)
    elif isinstance(raw, list):
        for v in raw:
            out |= _arg_deps(v)
    return out


def _toposort(defs, order, deps_of):
    pos = {n: i for i, n in enumerate(order)}
    pending = dict(defs)
    done, out = set(), []
    while pending:
        ready = sorted(
            (n for n, d in pending.items() if not (deps_of(d) & set(pending) - {n})),
            key=lambda n: pos.get(n, 0),
        )
        if not ready:
            raise DerivationError(f"dependency cycle among {sorted(pending)} (R001)")
        for n in ready:
            out.append(n)
            done.add(n)
            del pending[n]
    return out


def run_spec(path):
    return Spec(path).run()
